users/forms.py

```diff
@@ -1,30 +1,60 @@
-
-'''
-The MIT License (MIT)
-
-Copyright (c) 2014 NTHUOJ team
-
-Permission is hereby granted, free of charge, to any person obtaining a copy
-of this software and associated documentation files (the "Software"), to deal
-in the Software without restriction, including without limitation the rights
-to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-copies of the Software, and to permit persons to whom the Software is
-furnished to do so, subject to the following conditions:
-
-The above copyright notice and this permission notice shall be included in all
-copies or substantial portions of the Software.
-
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
-'''
 from django import forms
+from django.conf import settings
+from django.forms import ModelForm
 
 from users.models import User
+from problem.models import Problem, Submission, SubmissionDetail, Testcase
+from utils import log_info
+
+logger = log_info.get_logger()
+
+class CodeSubmitForm(forms.Form):
+    LANGUAGE_CHOICE = (
+        ('C', 'C: -O2 -lm -std=c99'),
+        ('CPP', 'C++:  -O2 -lm -std=c++'),
+        ('CPP11', 'C++11: -O2 -lm -std=c++11'),
+    )
+
+    pid = forms.CharField(
+        widget=forms.TextInput(attrs={'class': 'form-control', 'type': 'number'}))
+    language = forms.ChoiceField(choices=LANGUAGE_CHOICE, initial='CPP',
+        widget=forms.RadioSelect())
+    code = forms.CharField(max_length=10000,
+        widget=forms.Textarea(
+            attrs={'class': 'form-control', 'rows': 20,
+                'id': 'code_editor'}))
+
+    def clean_pid(self):
+        pid = self.cleaned_data['pid']
+        try:
+            Problem.objects.get(id=pid)
```

Comment:
Only admin, problem owner, contest(including the problem) owner / coowner can submit a invisible problem.